{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" \n",
    "Hot class 3 queries.\n",
    "\n",
    "First query (1.): Pattern matches for anything with word 'hot' in relevant comments boxes.\n",
    "Second query (2.): Find all classified class 3 variants, and subset those which went for Sanger validation.\n",
    "\n",
    "dbs copy:2019-09-09\n",
    "\"\"\"\n",
    "\n",
    "gelirs = GELInterpretationReport.objects.latest_cases_by_sample_type('raredisease')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Find hot class 3 case logic: \n",
    "\n",
    "1. All cases with 'hot' string in text fields                \n",
    "\"\"\"\n",
    "import re\n",
    "\n",
    "class_3_regex_cases = []\n",
    "\n",
    "for gelir in gelirs:\n",
    "    try:\n",
    "        if gelir.status != 'blocked':\n",
    "            gelir_id = gelir.ir_family\n",
    "            max_tier = gelir.max_tier\n",
    "            clinician = gelir.ir_family.participant_family.clinician.name\n",
    "            case_status = gelir.case_status\n",
    "\n",
    "            proband_comment = gelir.ir_family.participant_family.proband.comment\n",
    "            proband_discussion = gelir.ir_family.participant_family.proband.discussion\n",
    "            proband_action = gelir.ir_family.participant_family.proband.action\n",
    "\n",
    "            # Hot class 3 hunting, logic 1\n",
    "            all_text_fields = (proband_comment, proband_discussion, proband_action)\n",
    "            potential_hot_3 = re.search(r'(\\s+|\\'|\\.|\\\"|\\*)hot(\\s+|\\'|\\.|\\\"|\\*)', str(all_text_fields))\n",
    "\n",
    "            if potential_hot_3:\n",
    "                line = gelir.ir_family.participant_family.proband.gel_id, gelir.ir_family.ir_family_id\n",
    "                class_3_regex_cases.append(line)\n",
    "       \n",
    "    except Proband.DoesNotExist:\n",
    "        pass\n",
    "\n",
    "# Sanity check\n",
    "print(len(class_3_regex_cases))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output to xlsx\n",
    "import xlsxwriter\n",
    "\n",
    "workbook = xlsxwriter.Workbook(\"hot_class_3_cases_20190909.xlsx\")\n",
    "worksheet = workbook.add_worksheet('data')\n",
    "row = 0\n",
    "col = 0\n",
    "header = (\"gel_id\", \"cip-id\")\n",
    "worksheet.write_row(row, col, header)\n",
    "row = row + 1\n",
    "\n",
    "for report in class_3_regex_cases:\n",
    "    worksheet.write_row(row, col, report)\n",
    "    row = row + 1\n",
    "\n",
    "workbook.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Find hot class 3 case logic:\n",
    "\n",
    "2. All class 3 cases that have gone to sanger confirmation\n",
    "'''\n",
    "class_3_variants_all = []\n",
    "class_3_variants_validated = []\n",
    "\n",
    "for gelir in gelirs:\n",
    "    try:\n",
    "        if gelir.status != 'blocked':\n",
    "            if gelir.status != 'unknown':\n",
    "                gelir_cipid = gelir.ir_family\n",
    "                gelir_family_id = gelir.ir_family.ir_family_id\n",
    "                case_status = gelir.case_status\n",
    "\n",
    "                pvs = ProbandVariant.objects.filter(interpretation_report=gelir)\n",
    "\n",
    "                for pv in pvs:\n",
    "                    # If variant has gone to MDT\n",
    "                    try:\n",
    "                        raredisease_report = RareDiseaseReport.objects.get(proband_variant=pv)\n",
    "                        \n",
    "                        # If variant classified as class 3\n",
    "                        if raredisease_report.classification == '3':\n",
    "                            # try to get transcript variant data\n",
    "                            try:\n",
    "                                # get transcript\n",
    "                                ptv = pv.get_transcript_variant()\n",
    "                                hgvs_c = ptv.hgvs_c\n",
    "                                gene = str(ptv.transcript.gene)\n",
    "                            except:\n",
    "                                hgvs_c = \"\"\n",
    "                                gene = \"\"\n",
    "                            \n",
    "                            #12:g.1909905C>T\n",
    "                            #hgvs_g = str(pv.variant.chromosome) + \":g.\" + str(pv.variant.position) + str(pv.variant.reference) + \">\" + str(pv.variant.alternate)\n",
    "                            \n",
    "                            \n",
    "                            data = gelir_cipid.ir_family_id, pv.variant_id, str(pv.variant.genome_assembly), pv.variant.chromosome, pv.variant.position, pv.variant.reference, pv.variant.alternate, hgvs_c, gene, raredisease_report.classification, pv.get_validation_status_display()\n",
    "                            class_3_variants_all.append(data)\n",
    "\n",
    "                            # If variant is in validation process\n",
    "                            if pv.validation_status != 'U' and pv.validation_status != 'N':\n",
    "                                data = gelir_cipid.ir_family_id, pv.variant_id, raredisease_report.classification, pv.get_validation_status_display()\n",
    "                                class_3_variants_validated.append(data)\n",
    "\n",
    "                    except RareDiseaseReport.DoesNotExist:\n",
    "                        pass\n",
    "\n",
    "    except Proband.DoesNotExist:\n",
    "        pass\n",
    "\n",
    "# sanity check\n",
    "print(\"All class 3 variants={}\".format(len(class_3_variants_all)))\n",
    "print(\"Class 3 variants sent for validation={}\".format(len(class_3_variants_validated)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output to xlsx\n",
    "import xlsxwriter\n",
    "\n",
    "workbook = xlsxwriter.Workbook(\"class_3_variants_all_20190909.xlsx\")\n",
    "worksheet = workbook.add_worksheet('data')\n",
    "row = 0\n",
    "col = 0\n",
    "header = (\"cip_id\", \"variant_id\", \"build\", \"chrom\", \"pos\", \"ref\", \"alt\", \"hgvs_c\", \"gene\", \"variant_classification\", \"variant_status\")\n",
    "worksheet.write_row(row, col, header)\n",
    "row = row + 1\n",
    "\n",
    "for report in class_3_variants_all:\n",
    "    worksheet.write_row(row, col, report)\n",
    "    row = row + 1\n",
    "\n",
    "workbook.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Django Shell-Plus",
   "language": "python",
   "name": "django_extensions"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
